core/lib/zksync_core/src/consensus/storage.rs

```diff
@@ -0,0 +1,423 @@
+            storage
+                .put_block(ctx, genesis, operator_address)
+                .await
+                .wrap("put_block()")?;
+
+            // Find the last miniblock with consensus field set (aka head).
+            // We assume here that all blocks in range (genesis,head) also have consensus field set.
+            // WARNING: genesis should NEVER be moved to an earlier block.
+            storage
+                .find_head_number(ctx)
+                .await
+                .wrap("find_head_number()")?
+        };
+        Ok(Self {
+            genesis: genesis.header.number,
+            head: sync::watch::channel(head).0,
+            pool,
+            operator_address,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl WriteBlockStore for SignedBlockStore {
+    /// Verify that `payload` is a correct proposal for the block `block_number`.
+    async fn verify_payload(
+        &self,
+        ctx: &ctx::Ctx,
+        block_number: validator::BlockNumber,
+        payload: &validator::Payload,
+    ) -> ctx::Result<()> {
+        tracing::error!("SignedBlockStore::verify_payload({block_number:?})");
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        let want = storage
+            .fetch_payload(ctx, block_number, self.operator_address)
+            .await
+            .wrap("fetch_payload()")?
+            .context("unknown block")?;
+        if payload != &want.encode() {
+            return Err(anyhow::anyhow!("unexpected payload").into());
+        }
+        Ok(())
+    }
+
+    /// Puts a block into this storage.
+    async fn put_block(&self, ctx: &ctx::Ctx, block: &validator::FinalBlock) -> ctx::Result<()> {
+        tracing::error!("SignedBlockStore::put_block()");
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+
+        // Currently main node is the only validator, so it should be the only one creating new
+        // blocks. To ensure that no gaps in the blocks are created we check here that we always
+        // insert a new head.
+        let head = *self.head.borrow();
+        let head = storage
+            .find_head_forward(ctx, head, self.operator_address)
+            .await
+            .wrap("find_head_forward()")?;
+        if block.header.number != head.header.number.next() {
+            return Err(anyhow::anyhow!(
+                "expected block with number {}, got {}",
+                head.header.number.next(),
+                block.header.number
+            )
+            .into());
+        }
+
+        storage
+            .put_block(ctx, block, self.operator_address)
+            .await
+            .wrap("put_block()")
+    }
+}
+
+#[async_trait::async_trait]
+impl BlockStore for SignedBlockStore {
+    async fn head_block(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::FinalBlock> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        let head = *self.head.borrow();
```

Comment:
I decided to keep the semantics simple and delegate "head updates" solely to run_background_tasks. This will prevent not noticing a hard-to-debug situation in which one forgets to call run_background_tasks and all methods keep working efficiently, except for subscribe() which works usually-unless-we-stop-calling-other-methods.